### ISSUE TITLE
[RFC] Routing with Aurelia Documentation

### ISSUE THREAD

**Issue by jwx** [aurelia/aurelia#369]
# 💬 RFC
In addition to the "traditional url/path configured routing" the new router will have a viewport centric routing based on conventions. This RFC contains an early version of the first part of the "first contact document" for this non-traditional routing.
___
# Routing

Navigation in Aurelia is handled by viewports and components. In short, you put one or more viewports in your application and then you specify their content using links or code. Whenever someone navigates to a new ~~state of viewport components~~ route an entry will be added to the browser history, making it possible to use the browser's back, forward and refresh and get the right set of components. Let's take a look at an example. 

## Viewports

For the start view of our app, we're going to use two viewports:
```html
<!-- app.html -->
<body>
  <nav>
    <!-- We'll get back to this -->
  </nav>

  <viewport name="lists" used-by="authors,books" default="authors"></viewport>
  <viewport name="content" default="about"></viewport>
</body>
```
The first viewport we'll give the name _lists_ and specify that it's only used by the components/custom elements _Authors_ and _Books_. We also set _Authors_ as the default component, meaning that if nothing is specifically set for this viewport, it will display the _Authors_ component. For the second viewport we won't specify anything except giving it the name _content_ and setting the default to the _About_ component. Convention will then allow all components not used by another viewport to use it. Note: While it's not necessary to explicitly name a viewport (it will then be named _default_ by convention), it is recommended. Especially if you have more than one viewport in your application.

That's it for setting up the viewports. Now let's look at how to navigate and change the components in them.

## Navigation

Navigation can be triggered in html through `a`nchor tags or in code with the router's `goto` method. (There are more navigation methods available in the router, but we'll stick to `goto` for now.) A navigation segment in it's basic form consists of three parts: which _component_ to show, in which _viewport_ to show the component and what _parameters_ to pass into the component, like this
```html
"component@viewport(parameters)"
```
or this
```typescript
router.goto(Component, 'viewport', parameters);
```
But sometimes you don't want to pass _parameters_ to the _component_, so that part is optional. And if the router can deduce where a component should go by itself, for example based on the viewport settings we did earlier, _viewport_ can also be left out. So if you're using only one viewport, or have done a configuration similar to the one we did, you don't have to think about the viewport when navigating. Our `nav` part from above could then be
```html
  <nav>
    <ul>
      <li><a href="about">About</a></li>
      <li><a href="authors+about">Authors</a></li>
      <li><a href="books+about">Books</a></li>
    </ul>
  </nav>
```
What's happening here then? The `about` just specifies a component to navigate to, but what's the `+about` part after `authors` and `books`? Well, a navigation instruction can consist of several navigation segments, separated by `+`, meaning that we can change the components in more than one viewport in the same navigation instruction. The `authors+about` instruction is in its full form `authors@lists+about@content`, but since both `authors` and `books` are used by the `lists` viewport we don't have to specify viewport name for them. And since those two are the only components that can use the `lists` viewport, all other components, including `about`, has nowhere to go but the `content` viewport so we don't have to specify viewport for them either.

And that's it! We've now set up our main application navigation.

## Parameters

Quite often you'll want to pass parameters when navigating to a component. In our example, the _lists_ viewport might show the `Books` component which displays a list of books and if someone clicks on the title of a book, we want to display details about that book in our _content_ viewport.

So we'll have the `Books` component fetch a list of `books` from somewhere and then display it with this
```html
<ul>
  <li repeat.for="book of books"><a href="book(${book.id})">${book.title}</a></li>
</ul>
```
which could result in
```html
<ul>
  <li><a href="book(1)">The Colour of Magic</a></li>
  <li><a href="book(2)">Jingo</a></li>
  <li><a href="book(3)">Night Watch</a></li>
</ul>
```
The `href` attribute contains the navigation instruction and basically says "show the `Book` component where book is `1`". Since the _lists_ viewport is only available for the `Authors` and `Books` components, the `Book` component will be loaded into the only remaining viewport, _content_. When the router loads the `Book` component it will call a method called `enter` in it. This method will receive the parameter(s) specified in the `href` in a list as the first argument.
```typescript
enter(parameters) {
  id = parameters.length ? parameters[0] : null;
  // Fetch book data for id ...
}
```

### Named parameters

If we're not fond of unnamed parameters in a list, there's a simple way to get them named: add the name to the parameters in the `href`
```html
<a href="book(id=1)>The Colour of Magic"</a>
```
The first parameter to the `enter` method would now be an object:
```typescript
enter(parameters) {
  id = parameters ? parameters.id : null;
  // Fetch book data for id ...
}
```
It's also possible to pass more than one parameter to the `enter` method.
```html
<a href="book(id=1&language=sv)">Magins färg</a>
```
```typescript
enter(parameters) {
  id = parameters ? parameters.id : null;
  language = parameters ? parameters.language : 'en';
  // Fetch book data for id and language...
}
```
Now, however, our `href`s are getting a bit verbose due to the names. Fortunately, there'a another way to specify the parameter names: add a static property called `parameters` to the component:
```typescript
static parameters = ['id', 'language'];

enter(parameters) {
  id = parameters ? parameters.id : null;
  language = parameters ? parameters.language : 'en';
  // Fetch book data for id and language...
}
```
The `href` is once again just the values
```html
<a href="book(1&sv)">Magins färg</a>
```
Most of the cases, though, you'll probably only pass one parameter, some kind of `id`, so that the `href` will read "show component where [component name] is [id value]".

That's it for navigating around and passing parameters between components! 

Next we'll look into something that's a lot easier than it sounds: having components with their own viewports!

## Demo

An demo application according to the examples above can be found [here]( https://dist-nlbyng1ha.now.sh).

**Comment by jwx** [comment]
I should point out that the link syntax isn't set yet. One currently popular alternative is
```html
component!viewport(parameters)
```
turning the above links into
```html
<a href="about">About</a>
<a href="authors+about">Authors</a>
<a href="books+about">Books</a>

<a href="book(1)">The Colour of Magic</a>
<a href="book(2)">Jingo</a>
<a href="book(3)">Night Watch</a>

<a href="book(id=1)">The Colour of Magic"</a>
<a href="book(id=1,language=sv)">Magins färg"</a>
<a href="book(1,sv)">Magins färg"</a>
```
and here are some other links with the same syntax
```html
<a href="books!lists+book!content(id=1,mode=edit)">Book list and book 1 details</a>
<a href="books!lists+book!content(1,edit)">Book list and book 1 details</a>
<a href="books+book(id=1,mode=edit)">Book list and book 1 details</a>
<a href="books+book(1,edit)">Book list and book 1 details</a>
<a href="books!lists+book(id=1,mode=edit)">Book list and book 1 details</a>
<a href="books!lists+book(1,edit)">Book list and book 1 details</a>
<a href="books+book!content(id=1,mode=edit)">Book list and book 1 details</a>
<a href="books+book!content(1,edit)">Book list and book 1 details</a>
```
Any thoughts, preferences or suggestions regarding this is of course welcome!

**Comment by akircher** [comment]
Is there a point for a special syntax for named parameters?  It seems that an object would work just fine without learning a whole new "router syntax".  e.g.,

`book=id=1&language=sv` -> `book={id:1,language:sv}` or `book({id:1, language:sv})`

Aurelia already has a [binding syntax for css styles](https://aurelia.io/docs/binding/class-and-style#class) that might also work here `book='id:1; language:sv';`

Also could you share, for example, we the default is using `+` rather than the `&` so that the url looks more like a traditional url query string?  

This is great work and I am excited to use the router.  I am sure you have thought a lot about these and other syntax options and decided to go with something more unique and novel.  Unique is good if the semantics are much different than what more traditional syntax provides, but its unclear to me right now how the semantics are different.  This very well could be because because I am missing something so please let me know if that is the case.

**Comment by EisenbergEffect** [comment]
Thanks for the feedback @akircher This is very helpful. @jwx can you explore some of these ideas? In general, I prefer to lean away from micro-syntaxes, as they make learning a framework more difficult (and remembering it too). Perhaps one of these ideas can be used to smooth out the learning curve.

**Comment by fkleuver** [comment]
@jwx and I have discussed this fairly extensively, here's the bottom line (and to clarify/emphasize a few points):
- Syntax for configured (vCurrent-style) routing won't change. These are convention-based (unconfigured) routes, which is why they need a little bit of extra expression to tell the router what's what
- The current syntax is a fairly arbitrary pick, just something that makes it easy to test and develop (it is configurable)
- The syntax needs to be such that they are also valid URLs - entering them in the navigation bar should work without having to encode them. This leaves `$-_.+!*'(),` as the only candidates for symbols.
- A single path segment needs at least 1 piece of information (the component name) and can have at most 4 pieces of information per component (the scope name, component name, viewport namd and parameters). This means we need give or take 3 symbols (might be able to normalize it down to 2) in addition to`/` to put information in a segment that is not part of the segment itself.
- @jwx initial idea for this was `scope/component@viewport=parameters+scope/component@viewport=parameters`
- Other proposal is `scope/component@viewport(parameters)+scope/component@viewport(parameters)`
- Yet another proposal is `scope/component!viewport(parameters)+scope/component!viewport(parameters)` (you get the idea)

This is what we need to flesh out in a way that makes it as intuitive as possible.

**Comment by EisenbergEffect** [comment]
@fkleuver Note that more syntax options would be available, including object syntax, if we used a custom attribute instead of `href` directly. We should think about how easy it is to learn and remember the above. It is a form of configuration, even in an attempt to create a conventional router. The config has just been moved elsewhere and embedded within a micro syntax. If that syntax is difficult to learn/remember, then we'd end up with a "conventional" system that is worse than a configuration model.

Bottom line, we need to continue to think through this and try different things to find the right balance.

**Comment by fkleuver** [comment]
@EisenbergEffect You can't put braces in the address bar without encoding them.

**Comment by fkleuver** [comment]
To the best of my knowledge, the only thing that truly needs to be distinguishable from segments (something the route recognizer can do without special syntax) is parameters. But even that might be possible to infer with some clever logic. @jwx is trying to keep the code simple. It won't stay simple forever :)

I personally kind of like the idea of parenthesized parameters for convention-based routing. Specifying scope and viewport should be optional for advanced scenarios, so I imagine in the vast majority of cases you would just see urls like `foo/bar(5)/baz/qux(new)` and such

**Comment by EisenbergEffect** [comment]
I do like the parenthesis as well. Seems fairly intuitive.

My point from earlier is that we don't need to set a raw `href` attribute necessarily. Using a custom attribute, or href with a binding and value converter would enable other options for syntax while still enabling a the `href` to have a legal value. It might turn out that that is more complicated and even harder to remember, but I want to make sure we're exploring a variety of options so that we can get the best balance.

**Comment by blackgwe** [comment]
Great work! But difficult to understand, please remove sentences like 

> The _About_ is easy enough to understand. 

If the viewport is unspecified, the  components/custom element _About_ should be loaded into the default viewPort (the viewPort with name 'default'). But there is no viewport with the name 'default' and the 'default' parameter in the viewport means, that if nothing is specifically set for this viewport, it will display the _About_ component. This  really not easy to understand.

**Comment by fkleuver** [comment]
@EisenbergEffect Yeah, a point @jwx has been stressing as well is that the new router isn't going to need urls in order to work in the first place. The router is simply "thing that loads components into viewports".
Urls are one possible way to tell it what to load and where, but you can also just directly give it data or a component to load, and there would be nothing happening in the address bar (and thus there would be no limitations on the format with which you can control that). It's a slight paradigm shift.

The challenge here is to find the sweet spot of which subset of functionality convention-based routing should support in a way that makes it powerful enough to fit most use cases without creating a complicated new microsyntax to learn.

On the other hand, configuration-free routing could be powerful enough as a concept to justify the cost of a small microsyntax. The reason being: you can configure your routes on the fly simply by passing in the correct url. Changing the way the router loads something means you don't have to change code or re-deploy the application. Just change the url. This is huge.

**Comment by blackgwe** [comment]
> `<a href="authors+about">Authors</a>`
> The _authors+about_ instruction is in its full form _authors@lists_ + _about@default_

There is no viewport named _default_, do you mean _about@content_ ?

Why is it necessary to target `"authors+about"`, when `about` is the default component of the _content_ viewport?

**Comment by blackgwe** [comment]
your proposed syntax:

`books+book=5+book-details=5`
`books+book=3+book-details@books-tabs=3`

I'd prefer web-form query string syntax https://en.wikipedia.org/wiki/Query_string#Web_forms . it's easy to transform, if all parameters are passed to viewport. My proposal

`books?id=5&view-ports=book,book-details` 
   => resolved to your syntax `books=5+book=5+book-details=5`
`books?id=3&view-ports=book-details,books-tabs`  
   => resolved to your syntax  `books=5+book=5+book-details=5`
`about` 
  => resolved to your syntax `about`

**Comment by jwx** [comment]
@akircher @EisenbergEffect 
As @fkleuver says, we've got a limited set of symbol candidates for the string syntax since they need to be okay for the URLs. The only thing we would actually need that string syntax is for deep links into an application and manual typing into the browser location field, but those are use cases that I feel should remain.

Having said that, a custom attribute that takes object syntax along the lines of 
```typescript
route-href="{ component: 'book', viewport: 'details', parameters: { id: 1 } }"
```
has been planned all along. So while the string syntax is necessary for certain use cases, it's never been my intention that you'd _need_ to actually learn it.

The parenthesis are currently the strongest contender for parameters.

@blackgwe 
Let me first start by apologizing! I wrote this document for myself before I even started working on the router and I think I wanted to keep an encouraging tone. Regardless, stating that something is easy or hard to understand is inappropriate in any kind of documentation.

Next I have to apologize again! As you've figured out, I'm using the name `default` when I mean `content`. When I actually created the demo application I gave the viewport a name and changed the code snippets, but forgot to update the text. It's been fixed.

Thanks for your syntax proposal. Unfortunately, `?` is a reserved symbol in URLs as it's the indicator for a query string. But while it probably won't be a part of the out-of-the-box syntax, it will be possible for you to take "full" control over the URL syntax. Take a look at [this demo](https://github.com/aurelia/aurelia/issues/367#issuecomment-452143600) to see one of the possible ways to do this.

It wouldn't be necessary to specify `+about` if all I wanted the link to do was have the _lists_ viewport show the `authors` component, but in this (somewhat cryptic) case, I also wanted the _content_ viewport to be "reset" to the `about` component by the link. You can find the demo application [here](https://dist-nlbyng1ha.now.sh).

And you can actually test the difference yourself. First click the _Books_ link and then click a book title. You're now seeing the `books` and `book` components (and `book-details`, but that's a viewport within the `book` component). Now manually change the fragment part of the URL to `#/authors` to change the list to the `authors` list but leave the `book` as it was. Hit browser back and then change to `#/authors+about` to change the content of both viewports.


Thank you all for sharing your thoughts and expressing your opinions regarding the new router! It's really energizing!

**Comment by akircher** [comment]
@blackgwe @jwx  I understand that `?` is technically reserved only for the query string boundary but I have never had a problem with it. I make/use urls like this all the time:
`domain.com/login?redirect=domain.com/account?password-reset=true`
My point isn't around `?` per-say -  I am more so wondering if there are any workarounds for the url and other constraints that you are mentioning.

**Comment by akircher** [comment]
Actually I am no longer sure that the `?` is reserved [RFC3986](https://www.ietf.org/rfc/rfc3986.txt)*

>  The query component contains non-hierarchical data that, along with
   data in the path component (Section 3.3), serves to identify a
   resource within the scope of the URI's scheme and naming authority
   (if any).  ***The query component is indicated by the first question
   mark ("?") character and terminated by a number sign ("#") character
   or by the end of the URI. The characters slash ("/") and question mark ("?") may represent data within the query component.***  Beware that some older, erroneous
   implementations may not handle such data correctly when it is used as
   the base URI for relative references (Section 5.1), apparently
   because they fail to distinguish query data from path data when
   looking for hierarchical separators.  However, as query components
   are often used to carry identifying information in the form of
   "key=value" pairs and one frequently used value is a reference to
   another URI, it is sometimes better for usability to avoid percent-
   encoding those characters.

\* not sure if this is the latest spec or not

**Comment by davismj** [comment]
I like some of the ideas that I've heard here, but I'm still of the opinion that we're trying to load too much functionality on the convention, where configuration would be the better tool for the job and keep things clean and intuitive. 

I've taken the ideas from this proposal and expanded them into an alternative proposal here: https://github.com/aurelia/aurelia/issues/424. In this proposal I've pinned down two variables, viewport and parameters, to make the syntax quite a bit less complicated and more familiar.

**Comment by jwx** [comment]
@davismj 
What tools are better for the job and what clean and intuitive is is of course also very subjective. In my opinion, if I, in the simplest of cases, want to go to the component `Home` (or `HomeCustomElement` or `HomeRoute`) just doing any of 
```html
<a href="home">Home<a>
```
```typescript
router.goto(Home);
router.goto('home');
```
and no route table configuration at all is very clean and intuitive and not in any way bested by a route table configuration approach.

Just to get this straight, by making an _alternative_ proposal your opinion is that we shouldn't implement this proposal, or anything from https://github.com/aurelia/aurelia/issues/307 and https://github.com/aurelia/aurelia/issues/367, at all?

**Comment by davismj** [comment]
> and no route table configuration at all is very clean and intuitive and not in any way bested by a route table configuration approach.

Agree. That part matches up exactly with my proposal, and I used this RFC as inspiration. 

> Just to get this straight, by making an _alternative_ proposal your opinion is that we shouldn't implement this proposal, or anything from #307 and #367, at all?

It is an alternative because at some point it goes in a different direction and makes different assumptions, but mostly it is a superset of your proposals. Your proposals were inspiration and were included in this new one. 

The main difference is that I make assumptions (a) that there is one optional id parameters and (b) that the viewport is always the default viewport in the conventional case. Doing so allows me to greatly simplify the syntax into something much more familiar, with no need for special characters. In my experience working with lots of developers from lots of different backgrounds, (a) and (b) are true in the overwhelming majority of cases.

Take a look, especially at the last section.

**Comment by jwx** [comment]
I've of course read your proposal. I'm however not sure you've read, or understood, all of mine since you feel that your proposal is a superset of mine. Granted, this is probably due to me not being the best at explaining things in written English, so I can understand if there's some lack of clarity.

The special characters would only come into play in more complex cases. There are also a lot of cases more complex than the ones limited by your assumptions that still won't require special characters or a route table configuration. And, of course, using the API's `goto` method and `route-href` attribute the developer wouldn't have to work with the special characters at all but would work with objects instead. My proposal is more than a "special syntax backed by some conventions". I think it for many cases will be easier (very subjective, I know) to work with than route table configuration.

I might be wrong, of course. And there might be things in your proposal I don't understand (being rather locked in on my viewpoint by now). Could you perhaps show how the demo app in https://github.com/aurelia/aurelia/issues/367#issuecomment-455612499 would be configured/set up with your proposal? Since both proposals deal with the simple cases more or less the same, I think it'd be valuable to see any differences in the proposals in something that's at least a bit more complex.

I've hoped and believed from the beginning that the two proposals, dynamic routes and route table configuration, would co-exist in the end. It's discouraging to hear you now want to take that off the table.

**Comment by davismj** [comment]
> Could you perhaps show how the demo app in [#367 (comment)](https://github.com/aurelia/aurelia/issues/367#issuecomment-455612499) would be configured/set up with your proposal?

That's a great idea. Can you share the repo with me on discord so I can reuse the basic content, please? :)

**Comment by jwx** [comment]
It's an earlier version of https://github.com/aurelia/aurelia/tree/master/packages/router/test/e2e/doc-example, somewhere along https://github.com/aurelia/aurelia/commit/6fb5f9d1ad4be8fbc99857db880cbfd48e699c9d. Without looking too much into it I think the biggest difference is stateful viewports (and that's literally just the attribute `stateful` added to three `au-viewport`s), some of the texts and configuration options for delays. Basically, if you just clean the `au-viewport`s from everything except name (or even replace them with `router-view`s) you should have a pretty clean slate. Oh, and you won't need the `au-nav`/`addNav` either if you're using the RouteTable for nav data. 

I'm looking forward to seeing this.

While an injected RouteTable into applications is a good idea (and also mirrors the principle I have with `addNav`) I think pushing the dynamic routing into the route table is the wrong way to go. The principle I've had in mind from the beginning, and that already has the necessary extension points, is basically this:

1. The RouteTable (if it hasn't been deactivated) is checked for a match
a. if a match is found, it's translated into a set of `ViewportInstruction`s that's passed to the Router (3)
b. if match isn't found and dynamic routing isn't deactivated, the route is passed to dynamic routing (2)
2. Dynamic routing tries to resolve the route
a. if the route can be resolved, the set of `ViewportInstruction`s is passed to the Router (3)
b. if the route can't be resolved, the navigation failed. This in turn might trigger a new navigation or an error or custom behavior (leaving this for now)
3. The Router tries to execute the set of `ViewportInstruction`s
a. if all `ViewportInstructions`s are executed successfully, all is well and the navigation completes
b if not all `ViewportInstructions`s can be executed successfully, the navigation is cancelled and appropriate "no success actions" are triggered (leaving this too for now)

There are many advantages with this approach, but the "free mix and match" is probably one of the major ones. Especially since it will be possible for those developers who want to just skip 1. and 2. and talk directly to the Router with `ViewportInstruction`s. Another advantage is that the compositional viewport functionality that I've done for dynamic routing would be available for the RouteTable routes as well.

**Comment by davismj** [comment]
Adding dynamic routes to the RouteTable is exactly the kind of approach @fkleuver was outlining here: https://github.com/aurelia/aurelia/issues/307#issuecomment-457887949. That would merge step 1 and 2 above.

**Comment by jwx** [comment]
It's quite some time ago since I presented this approach to @fkleuver and just a few days ago we were still in agreement. His `The configured route logic will come first and detect whether a string is a known route and if not (depending on the router's settings) defer to the dynamic routing.` is fully in line with the above.

**Comment by ShawnTalbert** [comment]
is this router ready to start using or should I hold off?

**Comment by jwx** [comment]
For direct routing I think you can at least take it for a test drive. (But remember: pre-alpha!) Some live coded test driving can be found [here](https://www.youtube.com/playlist?list=PLWhGw6UrvSU5l3jY_89_ainf9Ofxy39FF).

**Comment by ShawnTalbert** [comment]
thanks @jwx , using your videos as a guide.

**Comment by Vheissu** [comment]
@jwx Is this issue still relevant or can it be closed?

**Comment by Sayan751** [comment]
@Vheissu @jwx I am closing this one, considering the number of related PRs. Feel free to reopen, in case of any residual work.
